src/core/thread/mle.cpp

```diff
@@ -246,6 +252,7 @@ ThreadError Mle::Stop(void)
         mNetif.RemoveUnicastAddress(mLeaderAloc);
     }
 
+    mNetif.GetPendingDataset().StopDelayTimer();
```

Comment:
Yes, this case is a problem, we need to restart the Delay Timer if it attach successfully without refreshing network Pending dataset.

I have added the logic to check Delay Timer state once it attach/reattach/re-sync to the network (call `IsDelayTimerRunning()` in `setStateChild()` and `setStateRouter()`), and restart the Delay Timer if it is necessary. Please have a review again, thanks.